### PR TITLE
Adding permission check when querying content using GraphQL

### LINKED ISSUE #12693: GraphQL does not respect permissions when querying contents
### Is your feature request related to a problem? Please describe.
I want to grant a user permissions to be able to query their own content only using GraphQL. However, there seems to be no way for me to do that like we can using the UI.

### Describe the solution you'd like
It would be nice to be able to apply the permissions like we do in the UI for GraphQL.

### Describe alternatives you've considered
Nothing yet, any recommendation?

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Fix #12693 

Prior this PR, a user with permission to GraphQL can see all content regardless of their permissions. With the new added filter, we honor their permissions.

Also, `GraphQLContentOptions.ConfigureContentType(...)` had no effect. Like if the user configure contentType as hidden, we still show it anyway. This was also fixed.

### CODE REVIEW

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs [hunk 1]
```diff
@@ -106,7 +106,7 @@ public bool IsCollapsedByDefault(ContentTypePartDefinition definition)
                 return true;
             }
 
-            var contentTypeOption = ContentTypeOptions.FirstOrDefault(ctp => ctp.ContentType == contentType);
+            var contentTypeOption = ContentTypeOptions.FirstOrDefault(ctp => String.Equals(ctp.ContentType, contentType, StringComparison.OrdinalIgnoreCase));
```
Comment (jtkech) [inline src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs:109]: Why are you adding so often `StringComparison.OrdinalIgnoreCase`, here better to keep `ctp.ContentType == contentType` unless it really fixes an issue.

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs:109]: @jtkech retrieving content types/part is case insensitive. And in this case, one can configure the content type with case insensitive. I can change it back if needed

Comment (jtkech) [reply src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs:109]: Oh yes, you did it recently for lookup/searching in some places, right? But here I think options are populated with `contentType` coming from existing `ContentTypePartDefinition`.

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs:109]: Yes that was added recently. In here not it does not come from the definition. There is a method `ConfigureContentType(...)` that configure the content type is a free form which is why I added it here

Comment (jtkech) [reply src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Options/GraphQLContentOptions.cs:109]: Ah okay, cool then

**Comment by MikeAlhayek** [comment]
@jtkech are you able to help with the failing test case? 

The problem is that the role "Anonymous" has permission to view all contents. So when the test `ShouldNotReturnBlogsWithoutViewBlogContentPermission` runs, it fails because it inherits all claims assigned to "Anonymous". 

I think all these GraphQl tests should run with "Anonymous" and "Authenticated" that has NO "ViewContents" permissions assigned to them. Is there a better approach for this beside creating a recipe for GraphQL and maybe `GraphQLContext`?

**Comment by jtkech** [comment]
Okay, I think I found the issue, I will comment it.

**Comment by MikeAlhayek** [comment]
> Okay, I think I found the issue, I will comment it.

@jtkech is the issue different than what I described above?

**Comment by jtkech** [comment]
So, I saw that when removing in `ContentTypeQuery` this new added line the test works.

    query.RequirePermission(new Permission("ExecuteGraphQL"));

Then I looked at `RequiresPermissionValidationRule.AuthorizeNodePermissionAsync()` where it checks if there are more than one `permissions`, in that case it checks the following.

    if (authorizationResults.Any(x => !true))

Which I think is wrong, so I replaced it by

    if (authorizationResults.Any(x => !x))

Then the test was working because a validation error was added.

---

That said maybe you don't need to add the `ExecuteGraphQL` permission as it is checked in the `GraphQLMiddleware`.

**Comment by MikeAlhayek** [comment]
yeah that works but I still think in all the GraphQL tests we should remove ViewContents permission from the  "Anonymous" and "Authenticated" roles. These tests should be adding their permission is each test. the only reason what you did works is because you enforce 2 permissions not just the 1 as we had.

**Comment by jtkech** [comment]
> yeah that works

What did you do? Just removing `query.RequirePermission(new Permission("ExecuteGraphQL"));` or by applying the fix `if (authorizationResults.Any(x => !x))`.

> These tests should be adding their permission is each test

In this test it adds explicitly the `ExecuteGraphQL` then it fails (not the test) because it doesn't have the permission, which is intended to make the test successful.

>  the only reason what you did works is because you enforce 2 permissions not just the 1 as we had.

I don't understand

**Comment by jtkech** [comment]
About `ConfigureContentType()`, yes I saw it, when you said `in a free form` I thought it could be provided e.g. by an API call, but it's a configuration by code only, so I changed my mind about the case sensitivity.

I think that's okay to keep it case sensitive, to force people to provide by code the exact type name, as it is automatically done when using the generic `ConfigurePart<TContentPart>()`.

**Comment by jtkech** [comment]
> the only reason what you did works is because you enforce 2 permissions not just the 1 as we had.

Just to say that I did nothing, at the end just applied the mentioned fix.

**Comment by jtkech** [comment]
Hmm, just saw that it doesn't check if `permissions.Count() == 0` ;)

**Comment by MikeAlhayek** [comment]
@jtkech I changed that. but one of the tests is still failing

**Comment by jtkech** [comment]
Okay I'm going to look at it

**Comment by jtkech** [comment]
Okay looks like we need to reference the same static permission instance added in the test.

    GraphQLApi.Permissions.ExecuteGraphQL,

So I could make it working by using the following in `ContentTypeQuery`

    query.RequirePermission(Apis.GraphQL.Permissions.ExecuteGraphQL);

But needed to reference the `OC.Apis.GraphQL` module project, would need to move the Permissions to a core/abstractions project.

But it shows that our fix is okay as it still works when having more than one permission, then for now I suggest to not add the `RequirePermission(Apis.GraphQL.Permissions.ExecuteGraphQL)` as it is already checked in the `GraphQLMiddleware` line 62.

**Comment by MikeAlhayek** [comment]
@jtkech the tests are pass now. Let me know if you see something else before an approval.

**Review by jtkech — APPROVED** [review]

**Pull request merged**